src/core/net/udp6.cpp

```diff
@@ -293,7 +309,7 @@ otError Udp::Connect(SocketHandle &aSocket, const SockAddr &aSockAddr)
     }
 
 #if OPENTHREAD_CONFIG_PLATFORM_UDP_ENABLE
-    if (!IsMlePort(aSocket.mSockName.mPort))
+    if (ShouldUsePlatformUdp(aSocket))
```

Comment:
```suggestion
    if (aSocket.mHandle != nullptr)
```

Reply:
Actually we always create platform socket in `Udp::Open` no matter it's MLE or TMF or other ports.

This behavior was so before this PR.

Reply:
Can we change it in this pr?

Reply:
I suspect that we always creates OS socket because we don't know if platform udp is required in `Udp::Open` yet. 

If we only want to create OS socket when it's really necessary, we need to create OS socket later in `Udp::Bind`, in which we can determine whether or not to use platform UDP.

Thoughts?

Reply:
agree to move it to bind.

Reply:
updated:
- creates OS socket in `Udp::Bind`
- Use `aSocket.mHandle` when appropriate

@bukepo

Reply:
update:

**reverted the change**

the recent commit failed in multiple tests. Turnout it's more complex than we thought. `BindToNetif` might be called before `Bind` which would require the platform socket to be ready. 

I would prefer to refactor it (along with Backbone Router related code) in another PR since it's out of the scope of this PR.